torchdp/compute_dp-sdg_privacy.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python3
+"""
+Both `apply_dp_sgd_analysis()` and `get_privacy_spent()` functions
+ are based on Google's TF Privacy:
+ https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/analysis/compute_dp_sgd_privacy.py
+
+Command-line script for computing privacy of a model trained with DP-SGD.
+The script applies the RDP accountant to estimate privacy budget of an iterated
+ Sampled Gaussian Mechanism.
+
+Example:
+  compute_dp_sgd_privacy
+    --dataset-size=60000 \
+    --batch-size=256 \
+    --sigma=1.12 \
+    --epochs=60 \
+    --delta=1e-5
+The output states that DP-SGD with these parameters satisfies (2.92, 1e-5)-DP.
+"""
+import argparse
+import math
+from privacy_analysis import compute_rdp, get_privacy_spent
+
+
+def apply_dp_sgd_analysis(q, sigma, steps, orders, delta):
+    """Compute and print results of DP-SGD analysis.
+
+   Compute_rdp requires that sigma be the ratio of the standard deviation of
+   the Gaussian noise to the l2-sensitivity of the function to which it is
+   added. Hence, sigma here corresponds to the `noise_multiplier` parameter
+   in the DP-SGD implementation found in privacy.optimizers.dp_optimizer
+  """
+    rdp = compute_rdp(q, sigma, steps, orders)
+
+    #  Slight adaptation from TF version, in which
+    # `get_privacy_spent()` has one more arguments and one more element in
+    #  returned tuple, because it can also compute delta for a given epsilon
+    #  (and not only compute epsilon for a targeted delta).
+    eps, opt_order = get_privacy_spent(orders, rdp, delta=delta)
+
+    print(
+        "DP-SGD with\n\tsampling rate = {:.3g}% and\n\tnoise_multiplier = {}"
+        "\n\titerated over {} steps\n satisfies".format(100 * q, sigma, steps),
+        end=" ",
+    )
+    print(
+        "differential privacy with\n\teps = {:.3g} and\n\tdelta = {}.".format(
+            eps, delta
+        )
+    )
+    print("The optimal RDP order is {}.".format(opt_order))
+
+    if opt_order == max(orders) or opt_order == min(orders):
+        print(
+            "The privacy estimate is likely to be improved by expanding "
+            "the set of orders."
+        )
+
+    return eps, opt_order
+
+
+def compute_dp_sgd_privacy(n, batch_size, noise_multiplier, epochs, delta):
+    """Compute epsilon based on the given hyperparameters."""
+    q = batch_size / n  # q - the sampling ratio.
+    if q > 1:
+        raise ValueError("n must be larger than the batch size.")
+    orders = (
```

Comment:
Make this a command line arg as welll

Reply:
Please see my review below, I don't manage to do it here :(

Reply:
It's ready on my computer, but I'm a bit lost about how to send it here. Maybe I'm blocked because there is a pending commit for changing the file name (I don't know how to create a new commit) ?
If not, maybe could you please tell me how to make the modification, since they occur in several parts ? Please be indulgent with my clumsiness ;-)

Reply:
Should be done (see in an other conversation, sorry but I don't know how to do it here!)